exchanges/huobi/huobi_websocket.go

```diff
@@ -531,15 +335,156 @@ func (h *HUOBI) WsProcessOrderbook(update *WsDepth, symbol string) error {
 	var newOrderBook orderbook.Base
 	newOrderBook.Asks = asks
 	newOrderBook.Bids = bids
-	newOrderBook.Pair = p
+	newOrderBook.Pair = s.Pairs[0]
 	newOrderBook.Asset = asset.Spot
 	newOrderBook.Exchange = h.Name
 	newOrderBook.VerifyOrderbook = h.CanVerifyOrderbook
-	newOrderBook.LastUpdated = time.UnixMilli(update.Timestamp)
+	newOrderBook.LastUpdated = update.Timestamp.Time()
 
 	return h.Websocket.Orderbook.LoadSnapshot(&newOrderBook)
 }
 
+func (h *HUOBI) wsHandleMyOrdersMsg(s *subscription.Subscription, respRaw []byte) error {
+	var msg wsOrderUpdateMsg
+	if err := json.Unmarshal(respRaw, &msg); err != nil {
+		return err
+	}
+	o := msg.Data
+	p, err := h.CurrencyPairs.Match(o.Symbol, s.Asset)
+	if err != nil {
+		return err
+	}
+	d := &order.Detail{
+		ClientOrderID:   o.ClientOrderID,
+		Price:           o.Price,
+		Amount:          o.Size,
+		ExecutedAmount:  o.ExecutedAmount,
+		RemainingAmount: o.RemainingAmount,
+		Exchange:        h.Name,
+		Side:            o.Side,
+		AssetType:       s.Asset,
+		Pair:            p,
+	}
+	if o.OrderID != 0 {
+		d.OrderID = strconv.FormatInt(o.OrderID, 10)
+	}
+	switch o.EventType {
+	case "trigger", "deletion", "cancellation":
+		d.LastUpdated = time.Unix(o.LastActTime*1000, 0)
```

Comment:
This clips the time values I went through and converted the rest over to type.Time for you:

```diff 
diff --git a/exchanges/huobi/huobi_test.go b/exchanges/huobi/huobi_test.go
index e44088354..958452307 100644
--- a/exchanges/huobi/huobi_test.go
+++ b/exchanges/huobi/huobi_test.go
@@ -35,6 +35,7 @@ import (
        testsubs "github.com/thrasher-corp/gocryptotrader/internal/testing/subscriptions"
        mockws "github.com/thrasher-corp/gocryptotrader/internal/testing/websocket"
        "github.com/thrasher-corp/gocryptotrader/portfolio/withdraw"
+       "github.com/thrasher-corp/gocryptotrader/types"
 )

 // Please supply you own test keys here for due diligence testing.
@@ -1404,8 +1405,8 @@ func TestWsAccountUpdate(t *testing.T) {
        close(h.Websocket.DataHandler)
        require.Len(t, h.Websocket.DataHandler, 3, "Must see correct number of records")
        exp := []WsAccountUpdate{
-               {Currency: "btc", AccountID: 123456, Balance: 23.111, ChangeType: "transfer", AccountType: "trade", ChangeTime: 1568601800000, SeqNum: 1},
-               {Currency: "btc", AccountID: 33385, Available: 2028.69, ChangeType: "order.match", AccountType: "trade", ChangeTime: 1574393385167, SeqNum: 2},
+               {Currency: "btc", AccountID: 123456, Balance: 23.111, ChangeType: "transfer", AccountType: "trade", ChangeTime: types.Time(time.UnixMilli(1568601800000)), SeqNum: 1},
+               {Currency: "btc", AccountID: 33385, Available: 2028.69, ChangeType: "order.match", AccountType: "trade", ChangeTime: types.Time(time.UnixMilli(1574393385167)), SeqNum: 2},
                {Currency: "usdt", AccountID: 14884859, Available: 20.29388158, Balance: 20.29388158, AccountType: "trade", SeqNum: 3},
        }
        for _, e := range exp {
@@ -1436,7 +1437,7 @@ func TestWsOrderUpdate(t *testing.T) {
                        Status:        order.Rejected,
                        ClientOrderID: "test1",
                        AssetType:     asset.Spot,
-                       LastUpdated:   time.Unix(1583853365586000, 0),
+                       LastUpdated:   time.UnixMicro(1583853365586000),
                },
                {
                        Exchange:      h.Name,
@@ -1445,7 +1446,7 @@ func TestWsOrderUpdate(t *testing.T) {
                        Status:        order.Cancelled,
                        ClientOrderID: "test2",
                        AssetType:     asset.Spot,
-                       LastUpdated:   time.Unix(1583853365586000, 0),
+                       LastUpdated:   time.UnixMicro(1583853365586000),
                },
                {
                        Exchange:      h.Name,
@@ -1458,7 +1459,7 @@ func TestWsOrderUpdate(t *testing.T) {
                        Amount:        2,
                        Type:          order.Limit,
                        OrderID:       "27163533",
-                       LastUpdated:   time.Unix(1583853365586000, 0),
+                       LastUpdated:   time.UnixMicro(1583853365586000),
                },
                {
                        Exchange:    h.Name,
@@ -1470,7 +1471,7 @@ func TestWsOrderUpdate(t *testing.T) {
                        Amount:      0.000157,
                        Type:        order.Limit,
                        OrderID:     "1199329381585359",
-                       LastUpdated: time.Unix(1731039387696000, 0),
+                       LastUpdated: time.UnixMicro(1731039387696000),
                },
        }
        for _, e := range exp {
@@ -1500,8 +1501,8 @@ func TestWsMyTrades(t *testing.T) {
                ClientOrderID: "a001",
                OrderID:       "99998888",
                AssetType:     asset.Spot,
-               Date:          time.Unix(1583853365586000, 0),
-               LastUpdated:   time.Unix(1583853365996000, 0),
+               Date:          time.UnixMicro(1583853365586000),
+               LastUpdated:   time.UnixMicro(1583853365996000),
                Price:         10000,
                Amount:        1,
                Trades: []order.TradeHistory{
@@ -1513,7 +1514,7 @@ func TestWsMyTrades(t *testing.T) {
                                TID:       "919219323232",
                                Side:      order.Buy,
                                IsMaker:   false,
-                               Timestamp: time.Unix(1583853365996000, 0),
+                               Timestamp: time.UnixMicro(1583853365996000),
                        },
                },
        }
@@ -2037,7 +2038,7 @@ func updatePairsOnce(tb testing.TB, h *HUOBI) {

        testexch.UpdatePairsOnce(tb, h)

-       if btcFutureDatedPair == currency.EMPTYPAIR {
+       if btcFutureDatedPair.Equal(currency.EMPTYPAIR) {
                p, err := h.pairFromContractExpiryCode(btccwPair)
                require.NoError(tb, err, "pairFromContractCode must not error")
                btcFutureDatedPair = p
diff --git a/exchanges/huobi/huobi_types.go b/exchanges/huobi/huobi_types.go
index b460f912e..80a04aaa1 100644
--- a/exchanges/huobi/huobi_types.go
+++ b/exchanges/huobi/huobi_types.go
@@ -905,14 +905,14 @@ type wsAccountUpdateMsg struct {

 // WsAccountUpdate contains account updates to balances
 type WsAccountUpdate struct {
-       Currency    string  `json:"currency"`
-       AccountID   int64   `json:"accountId"`
-       Balance     float64 `json:"balance,string"`
-       Available   float64 `json:"available,string"`
-       ChangeType  string  `json:"changeType"`
-       AccountType string  `json:"accountType"`
-       ChangeTime  int64   `json:"changeTime"`
-       SeqNum      int64   `json:"seqNum"`
+       Currency    string     `json:"currency"`
+       AccountID   int64      `json:"accountId"`
+       Balance     float64    `json:"balance,string"`
+       Available   float64    `json:"available,string"`
+       ChangeType  string     `json:"changeType"`
+       AccountType string     `json:"accountType"`
+       ChangeTime  types.Time `json:"changeTime"`
+       SeqNum      int64      `json:"seqNum"`
 }

 type wsOrderUpdateMsg struct {
@@ -939,9 +939,9 @@ type WsOrderUpdate struct {
        IsTaker         bool       `json:"aggressor"`
        Side            order.Side `json:"orderSide"`
        OrderStatus     string     `json:"orderStatus"`
-       LastActTime     int64      `json:"lastActTime"`
-       CreateTime      int64      `json:"orderCreateTime"`
-       TradeTime       int64      `json:"tradeTime"`
+       LastActTime     types.Time `json:"lastActTime"`
+       CreateTime      types.Time `json:"orderCreateTime"`
+       TradeTime       types.Time `json:"tradeTime"`
        ErrCode         int64      `json:"errCode"`
        ErrMessage      string     `json:"errMessage"`
 }
@@ -961,7 +961,7 @@ type WsTradeUpdate struct {
        OrderType       string     `json:"orderType"`
        IsTaker         bool       `json:"aggressor"`
        TradeID         int64      `json:"tradeId"`
-       TradeTime       int64      `json:"tradeTime"`
+       TradeTime       types.Time `json:"tradeTime"`
        TransactFee     float64    `json:"transactFee,string"`
        FeeCurrency     string     `json:"feeCurrency"`
        FeeDeduct       string     `json:"feeDeduct"`
@@ -974,7 +974,7 @@ type WsTradeUpdate struct {
        ClientOrderID   string     `json:"clientOrderId"`
        StopPrice       string     `json:"stopPrice"`
        Operator        string     `json:"operator"`
-       OrderCreateTime int64      `json:"orderCreateTime"`
+       OrderCreateTime types.Time `json:"orderCreateTime"`
        OrderStatus     string     `json:"orderStatus"`
 }

@@ -1169,6 +1169,6 @@ type WithdrawalData struct {
        State           string        `json:"state"`
        ErrorCode       string        `json:"error-code"`
        ErrorMessage    string        `json:"error-message"`
-       CreatedAt       int64         `json:"created-at"`
-       UpdatedAt       int64         `json:"updated-at"`
+       CreatedAt       types.Time    `json:"created-at"`
+       UpdatedAt       types.Time    `json:"updated-at"`
 }
diff --git a/exchanges/huobi/huobi_websocket.go b/exchanges/huobi/huobi_websocket.go
index 0df94c208..44611144b 100644
--- a/exchanges/huobi/huobi_websocket.go
+++ b/exchanges/huobi/huobi_websocket.go
@@ -370,11 +370,11 @@ func (h *HUOBI) wsHandleMyOrdersMsg(s *subscription.Subscription, respRaw []byte
        }
        switch o.EventType {
        case "trigger", "deletion", "cancellation":
-               d.LastUpdated = time.Unix(o.LastActTime*1000, 0)
+               d.LastUpdated = o.LastActTime.Time()
        case "creation":
-               d.LastUpdated = time.Unix(o.CreateTime*1000, 0)
+               d.LastUpdated = o.CreateTime.Time()
        case "trade":
-               d.LastUpdated = time.Unix(o.TradeTime*1000, 0)
+               d.LastUpdated = o.TradeTime.Time()
        }
        if d.Status, err = order.StringToOrderStatus(o.OrderStatus); err != nil {
                return &order.ClassificationError{
@@ -428,8 +428,8 @@ func (h *HUOBI) wsHandleMyTradesMsg(s *subscription.Subscription, respRaw []byte
                Side:          t.Side,
                AssetType:     s.Asset,
                Pair:          p,
-               Date:          time.Unix(t.OrderCreateTime*1000, 0),
-               LastUpdated:   time.Unix(t.TradeTime*1000, 0),
+               Date:          t.OrderCreateTime.Time(),
+               LastUpdated:   t.TradeTime.Time(),
                OrderID:       strconv.FormatInt(t.OrderID, 10),
        }
        if d.Status, err = order.StringToOrderStatus(t.OrderStatus); err != nil {
@@ -469,7 +469,7 @@ func (h *HUOBI) wsHandleMyTradesMsg(s *subscription.Subscription, respRaw []byte
                        Type:      d.Type,
                        Side:      d.Side,
                        IsMaker:   !t.IsTaker,
-                       Timestamp: time.Unix(t.TradeTime*1000, 0),
+                       Timestamp: t.TradeTime.Time(),
                },
        }
        h.Websocket.DataHandler <- d
diff --git a/exchanges/huobi/huobi_wrapper.go b/exchanges/huobi/huobi_wrapper.go
index 8ff734c4c..a27a16e79 100644
--- a/exchanges/huobi/huobi_wrapper.go
+++ b/exchanges/huobi/huobi_wrapper.go
@@ -877,7 +877,7 @@ func (h *HUOBI) GetWithdrawalsHistory(ctx context.Context, c currency.Code, a as
                resp[i] = exchange.WithdrawalHistory{
                        Status:          withdrawals.Data[i].State,
                        TransferID:      withdrawals.Data[i].TransactionHash,
-                       Timestamp:       time.UnixMilli(withdrawals.Data[i].CreatedAt),
+                       Timestamp:       withdrawals.Data[i].CreatedAt.Time(),
                        Currency:        withdrawals.Data[i].Currency.String(),
                        Amount:          withdrawals.Data[i].Amount,
                        Fee:             withdrawals.Data[i].Fee,
```

Reply:
Just realised I included `if btcFutureDatedPair.Equal(currency.EMPTYPAIR) {` you don't have to change that

Reply:
Fixed https://github.com/gbjk/gocryptotrader/commit/4d0a56c0b